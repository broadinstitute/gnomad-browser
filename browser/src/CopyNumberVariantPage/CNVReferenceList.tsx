import React from 'react'

import { ExternalLink, List, ListItem } from '@gnomad/ui'

import { CopyNumberVariant } from './CopyNumberVariantPage'

const ucscUrl = (chrom: any, pos: any, end: any) =>
  `https://genome.ucsc.edu/cgi-bin/hgTracks?db=hg38&position=chr${chrom}%3A${pos}-${end}`

type CNVUCSCLinksProps = {
  variant: CopyNumberVariant
}

const CNVUCSCLinks = ({ variant }: CNVUCSCLinksProps) => {
  // @ts-expect-error TS(2786) FIXME: 'ExternalLink' cannot be used as a JSX component.
  return <ExternalLink href={ucscUrl(variant.chrom, variant.pos, variant.end)}>UCSC</ExternalLink>
}

type CNVReferenceListProps = {
  variant: CopyNumberVariant
}

const CNVReferenceList = ({ variant }: CNVReferenceListProps) => (
  // @ts-expect-error TS(2745) FIXME: This JSX tag's 'children' prop expects type 'never... Remove this comment to see the full error message
  <List>
    {/* @ts-expect-error TS(2745) FIXME: This JSX tag's 'children' prop expects type 'never... Remove this comment to see the full error message */}
    <ListItem>
      <CNVUCSCLinks variant={variant} />
    </ListItem>
  </List>
)

export default CNVReferenceList
