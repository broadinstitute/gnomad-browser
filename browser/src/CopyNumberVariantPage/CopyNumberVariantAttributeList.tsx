import React from 'react'

import { Badge } from '@gnomad/ui'

import AttributeList, { AttributeListItem } from '../AttributeList'
import Link from '../Link'
import { cnvTypeLabels } from '../CopyNumberVariantList/copyNumberVariantTypes'
import { CopyNumberVariant } from './CopyNumberVariantPage'
import { textOrMissingTextWarning } from '../missingContent'

const FILTER_LABELS = {
  TRUE: 'TRUE',
}

const FILTER_DESCRIPTIONS = {
  TRUE: 'true',
}

const filterLabel = (filter: string) =>
  textOrMissingTextWarning('filter label', FILTER_LABELS, filter)

const filterDescription = (filter: string) =>
  textOrMissingTextWarning('filter description', FILTER_DESCRIPTIONS, filter)

type CopyNumberVariantAttributeListProps = {
  variant: CopyNumberVariant
}

const CopyNumberVariantAttributeList = ({ variant }: CopyNumberVariantAttributeListProps) => (
  <AttributeList style={{ marginTop: '1.25em' }}>
    <AttributeListItem label="Filter">
      {variant.filters.length > 0 ? (
        variant.filters.map((filter) => (
          <Badge key={filter} level="warning" tooltip={filterDescription(filter)}>
            {filterLabel(filter)}
          </Badge>
        ))
      ) : (
        <Badge level="success">Pass</Badge>
      )}
    </AttributeListItem>

    <AttributeListItem label="Site Count">{variant.sc}</AttributeListItem>
    <AttributeListItem label="Site Number">{variant.sn}</AttributeListItem>
    <AttributeListItem label="Site Frequency">
      {(variant.sn === 0 ? 0 : variant.sc / variant.sn).toPrecision(4)}
    </AttributeListItem>
    <AttributeListItem label="Position">
      <Link to={`/region/${variant.chrom}-${variant.pos}-${variant.end}`}>
        {variant.chrom}:{variant.pos}-{variant.end}
      </Link>
    </AttributeListItem>
    <AttributeListItem label="Size">
      {variant.length === -1 ? 'â€”' : `${variant.length.toLocaleString()} bp`}
    </AttributeListItem>
    <AttributeListItem label="Class">
      {/* @ts-expect-error TS(7053) FIXME: Element implicitly has an 'any' type because expre... Remove this comment to see the full error message */}
      {cnvTypeLabels[variant.type]}
    </AttributeListItem>
  </AttributeList>
)

export default CopyNumberVariantAttributeList
