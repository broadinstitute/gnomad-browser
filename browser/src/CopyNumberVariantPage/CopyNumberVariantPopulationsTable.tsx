import React from 'react'

import { CopyNumberVariant } from './CopyNumberVariantPage'
import { populationName } from '@gnomad/dataset-metadata/gnomadPopulations'
import { CNVPopulationsTable } from './CNVPopulationsTable'

type CopyNumberVariantPopulation = {
  id: string
  sc: number
  sn: number
  sf: number
  name: string
}

type Subpopulations = {
  [key: string]: CopyNumberVariantPopulation[]
}

const nestPopulations = (populations: CopyNumberVariantPopulation[]) => {
  const popIndices = []
  const subpopulations: Subpopulations = {}

  for (let i = 0; i < populations.length; i += 1) {
    const pop = populations[i]

    const divisions = pop.id.split('_')
    if (divisions.length === 1) {
      popIndices.push(i)
    } else {
      const parentPop = divisions[0]
      if (subpopulations[parentPop] === undefined) {
        subpopulations[parentPop] = [{ ...pop }]
      } else {
        subpopulations[parentPop].push({ ...pop })
      }
    }
  }

  return popIndices.map((index) => {
    const pop = populations[index]
    return {
      ...pop,
      subpopulations: subpopulations[pop.id],
    }
  })
}

const addPopulationNames = (populations: CopyNumberVariantPopulation[]) => {
  return populations.map((pop: CopyNumberVariantPopulation) => {
    let name
    if (pop.id === 'XX' || pop.id.endsWith('_XX')) {
      name = 'XX'
    } else if (pop.id === 'XY' || pop.id.endsWith('_XY')) {
      name = 'XY'
    } else {
      name = populationName(pop.id)
    }
    return { ...pop, name }
  })
}

type CopyNumberVariantPopulationsTableProps = {
  variant: CopyNumberVariant
}

const CopyNumberVariantPopulationsTable = ({ variant }: CopyNumberVariantPopulationsTableProps) => {
  const populations = nestPopulations(addPopulationNames((variant as any).populations))

  const columnLabels = {
    sc: 'Site Count',
    sn: 'Site Number',
    sf: 'Site Frequency',
  }

  return <CNVPopulationsTable columnLabels={columnLabels} populations={populations} variant={variant} />
}

export default CopyNumberVariantPopulationsTable
