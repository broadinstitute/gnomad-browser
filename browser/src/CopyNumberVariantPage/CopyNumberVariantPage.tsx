import React from 'react'
import styled from 'styled-components'

import { ExternalLink, Page, ListItem } from '@gnomad/ui'
import Link from '../Link'

import { DatasetId, labelForDataset } from '@gnomad/dataset-metadata/metadata'
import DocumentTitle from '../DocumentTitle'
import GnomadPageHeading from '../GnomadPageHeading'
import Query from '../Query'
import { variantFeedbackUrl } from '../variantFeedback'
import CopyNumberVariantAttributeList from './CopyNumberVariantAttributeList'
import CopyNumberVariantPopulationsTable from './CopyNumberVariantPopulationsTable'
import CNVReferenceList from './CNVReferenceList'

const Wrapper = styled.div`
  display: flex;
  flex-flow: row wrap;
  justify-content: space-between;
`

const ResponsiveSection = styled.section`
  width: calc(50% - 15px);

  @media (max-width: 992px) {
    width: 100%;
  }
`

export type CopyNumberVariant = {
  alts: string[] | null
  sc: number
  sn: number
  sf: number
  chrom: string
  end: number
  filters: string[]
  genes: string[]
  length: number
  populations: {
    id: string
    sc: number
    sn: number
    sf: number
  }[]
  pos: number
  qual: number
  type: string
  posmin: number
  posmax: number
  endmin: number
  endmax: number
  variant_id: string
}

type CopyNumberVariantPageProps = {
  datasetId: DatasetId
  variant: CopyNumberVariant
}

const CopyNumberVariantPage = ({ datasetId, variant }: CopyNumberVariantPageProps) => (
  // @ts-expect-error TS(2746) FIXME: This JSX tag's 'children' prop expects a single ch... Remove this comment to see the full error message
  <Page>
    <DocumentTitle title={`${variant.variant_id} | ${labelForDataset(datasetId)}`} />
    <GnomadPageHeading
      datasetOptions={{ includeShortVariants: false, includeStructuralVariants: false }}
      selectedDataset={datasetId}
    >
      Copy number variant: {variant.variant_id}
    </GnomadPageHeading>
    <Wrapper>
      <ResponsiveSection>
        <CopyNumberVariantAttributeList variant={variant} />
      </ResponsiveSection>
      <ResponsiveSection>
        <h2>External Resources</h2>
        <CNVReferenceList variant={variant} />
        <h2>Feedback</h2>
        {/* @ts-expect-error TS(2786) FIXME: 'ExternalLink' cannot be used as a JSX component. */}
        <ExternalLink href={variantFeedbackUrl(variant, datasetId)}>
          Report an issue with this variant
        </ExternalLink>
      </ResponsiveSection>
    </Wrapper>

    <section>
      <h2>Population Frequencies</h2>
      <CopyNumberVariantPopulationsTable variant={variant} />
    </section>

    <Wrapper>
      <ResponsiveSection>
        <h2>Consequences</h2>
        <p>
          This variant has consequences in {variant.genes.length} gene
          {variant.genes.length !== 1 && 's'}.
        </p>
        {variant.genes.map((gene) => (
          // @ts-expect-error TS(2769) FIXME: No overload matches this call.
          <ListItem key={gene}>
            <Link to={`/gene/${gene}`}>{gene}</Link>
          </ListItem>
        ))}
      </ResponsiveSection>
    </Wrapper>
  </Page>
)

type ConnectedCopyNumberVariantPageProps = {
  datasetId: DatasetId
  variantId: string
}

const ConnectedCopyNumberVariantPage = ({
  datasetId,
  variantId,
}: ConnectedCopyNumberVariantPageProps) => {
  const operationName = 'CopyNumberVariant'

  const query = `
    query ${operationName}($datasetId: CopyNumberVariantDatasetId!, $variantId: String!) {
      copy_number_variant(dataset: $datasetId, variantId: $variantId) {
        alts
        sc
        sn
        sf
        chrom
        end
        filters
        genes
        length
        populations {
            id
            sc
            sn
            sf
        }
        pos
        qual
        reference_genome
        type
        posmin
        posmax
        endmin
        endmax
        variant_id
      }
    }
  `

  return (
    <Query
      operationName={operationName}
      query={query}
      variables={{ datasetId, variantId }}
      loadingMessage="Loading variant"
      errorMessage="Unable to load variant"
      success={(data: any) => data.copy_number_variant}
    >
      {({ data }: any) => {
        const variant = {
          ...data.copy_number_variant,
          variant_id: data.copy_number_variant.variant_id,
        }
        return <CopyNumberVariantPage datasetId={datasetId} variant={variant} />
      }}
    </Query>
  )
}

export default ConnectedCopyNumberVariantPage
