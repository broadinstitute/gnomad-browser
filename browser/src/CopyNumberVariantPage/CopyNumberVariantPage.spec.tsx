import React from 'react'
import renderer from 'react-test-renderer'
import { jest, expect, test, describe } from '@jest/globals'
import { mockQueries } from '../../../tests/__helpers__/queries'
import Query, { BaseQuery } from '../Query'
import { forDatasetsMatching } from '../../../tests/__helpers__/datasets'
import CopyNumberVariantPage from './CopyNumberVariantPage'
import cnvFactory from '../__factories__/CopyNumberVariant'
import { BrowserRouter } from 'react-router-dom'

jest.mock('../Query', () => {
  const originalModule = jest.requireActual('../Query')

  return {
    __esModule: true,
    ...(originalModule as object),
    default: jest.fn(),
    BaseQuery: jest.fn(),
  }
})

const { resetMockApiCalls, resetMockApiResponses, simulateApiResponse, setMockApiResponses } =
  mockQueries()

beforeEach(() => {
  Query.mockImplementation(
    jest.fn(({ query, children, operationName, variables }) =>
      simulateApiResponse('Query', query, children, operationName, variables)
    )
  )
  ;(BaseQuery as any).mockImplementation(
    jest.fn(({ query, children, operationName, variables }) =>
      simulateApiResponse('BaseQuery', query, children, operationName, variables)
    )
  )
})

afterEach(() => {
  resetMockApiCalls()
  resetMockApiResponses()
})

forDatasetsMatching(/gnomad_cnv_r4/, 'CopyNumberVariantPage with dataset %s', (datasetId) => {
  describe.each(['DEL', 'DUP'])('with variant of type %s', (variantType: string) => {
    test('has no unexpected changes', () => {
      const variant = cnvFactory.build({ type: variantType })
      setMockApiResponses({
        CopyNumberVariant: () => ({
          copy_number_variant: variant,
        }),
      })
      const tree = renderer.create(
        <BrowserRouter>
          <CopyNumberVariantPage datasetId={datasetId} variantId={variant.variant_id} />
        </BrowserRouter>
      )
      expect(tree).toMatchSnapshot()
    })
  })
})
