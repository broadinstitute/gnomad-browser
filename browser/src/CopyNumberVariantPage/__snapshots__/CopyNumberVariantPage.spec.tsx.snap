// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CopyNumberVariantPage with dataset gnomad_cnv_r4 with variant of type DEL has no unexpected changes 1`] = `
<div
  className="Page-sc-w7h773-0 kodnkv"
>
  <div
    className="GnomadPageHeading__PageHeadingWrapper-sc-sgz3rr-0 gcOcdP"
  >
    <div
      className="GnomadPageHeading__PageHeadingInnerWrapper-sc-sgz3rr-1 dLZzAa"
    >
      <h1
        className="GnomadPageHeading__PageHeadingText-sc-sgz3rr-3 edXBqX"
      >
        Copy number variant: 
        21__DEL
      </h1>
    </div>
    <div
      className="GnomadPageHeading__PageControlsWrapper-sc-sgz3rr-4 liqJpb"
    >
      <span
        className="GnomadPageHeading__Label-sc-sgz3rr-5 dOhShJ"
      >
        Dataset
      </span>
      <ul
        className="DatasetSelector__NavigationMenuWrapper-sc-1p2fxbn-0 ebZvwy"
      >
        <li
          className="DatasetSelector__NavigationMenuItem-sc-1p2fxbn-2 ifirRc"
        >
          <a
            className="DatasetSelector__TopLevelNavigationLink-sc-1p2fxbn-1-Link eggTrD"
            data-item="current_sv_dataset"
            href="/?dataset=gnomad_cnv_r4"
            onBlur={[Function]}
            onClick={[Function]}
            onKeyDown={[Function]}
          >
            gnomAD CNVs v4.0
          </a>
        </li>
        <li
          className="DatasetSelector__NavigationMenuItem-sc-1p2fxbn-2 ifirRc"
        >
          <a
            aria-expanded={false}
            aria-haspopup="true"
            aria-label="More datasets"
            className="DatasetSelector__TopLevelNavigationLink-sc-1p2fxbn-1-Link eggTrD"
            data-item="other_structural_variant"
            href="/"
            onBlur={[Function]}
            onClick={[Function]}
            onKeyDown={[Function]}
            role="button"
          >
            <span
              className="DatasetSelector__MoreIcon-sc-1p2fxbn-3 hTYFQu"
            >
              <img
                alt=""
                aria-hidden="true"
                src="test-file-stub"
              />
            </span>
          </a>
          <ul
            className="DatasetSelector__SubNavigationMenu-sc-1p2fxbn-4 epYyTQ"
          >
            <li>
              <div
                className="DatasetSelector__GroupedNav-sc-1p2fxbn-7 cJJZPz"
              >
                GRCh38
              </div>
              <li>
                <a
                  className="DatasetSelector__SubNavigationLink-sc-1p2fxbn-5-Link gbnrKe"
                  data-item="gnomad_sv_r4"
                  href="/?dataset=gnomad_sv_r4"
                  onBlur={[Function]}
                  onClick={[Function]}
                  onKeyDown={[Function]}
                >
                  gnomAD SVs v4.0
                  <div
                    className="DatasetSelector__ItemDescription-sc-1p2fxbn-6 jjgqnf"
                  >
                    63,046 samples, genome
                  </div>
                </a>
              </li>
              <li>
                <a
                  className="DatasetSelector__SubNavigationLink-sc-1p2fxbn-5-Link gbnrKe"
                  data-item="gnomad_cnv_r4"
                  href="/?dataset=gnomad_cnv_r4"
                  onBlur={[Function]}
                  onClick={[Function]}
                  onKeyDown={[Function]}
                >
                  gnomAD CNVs v4.0
                  <div
                    className="DatasetSelector__ItemDescription-sc-1p2fxbn-6 jjgqnf"
                  >
                    464,297 samples, exome, rare (&lt;0.01)
                  </div>
                </a>
              </li>
            </li>
            <li>
              <div
                className="DatasetSelector__GroupedNav-sc-1p2fxbn-7 cJJZPz"
              >
                GRCh37
              </div>
              <li>
                <a
                  className="DatasetSelector__SubNavigationLink-sc-1p2fxbn-5-Link gbnrKe"
                  data-item="gnomad_sv_r2_1"
                  href="/?dataset=gnomad_sv_r2_1"
                  onBlur={[Function]}
                  onClick={[Function]}
                  onKeyDown={[Function]}
                >
                  gnomAD SVs v2.1
                  <div
                    className="DatasetSelector__ItemDescription-sc-1p2fxbn-6 jjgqnf"
                  >
                    10,847 samples, genome
                  </div>
                </a>
              </li>
              <li>
                <a
                  className="DatasetSelector__SubNavigationLink-sc-1p2fxbn-5-Link gbnrKe"
                  data-item="gnomad_sv_r2_1_non_neuro"
                  href="/?dataset=gnomad_sv_r2_1_non_neuro"
                  onBlur={[Function]}
                  onClick={[Function]}
                  onKeyDown={[Function]}
                >
                  gnomAD SVs v2.1 (non-neuro)
                  <div
                    className="DatasetSelector__ItemDescription-sc-1p2fxbn-6 jjgqnf"
                  >
                    8,342 samples, genome
                  </div>
                </a>
              </li>
              <li>
                <a
                  className="DatasetSelector__SubNavigationLink-sc-1p2fxbn-5-Link gbnrKe"
                  data-item="gnomad_sv_r2_1_controls"
                  href="/?dataset=gnomad_sv_r2_1_controls"
                  onBlur={[Function]}
                  onClick={[Function]}
                  onKeyDown={[Function]}
                >
                  gnomAD SVs v2.1 (controls)
                  <div
                    className="DatasetSelector__ItemDescription-sc-1p2fxbn-6 jjgqnf"
                  >
                    5,192 samples, genome
                  </div>
                </a>
              </li>
            </li>
          </ul>
        </li>
      </ul>
      <span>
        <button
          className="InfoButton__Button-sc-13t5e82-0 gxsoyZ"
          onClick={[Function]}
          type="button"
        >
          <img
            alt=""
            aria-hidden="true"
            src="test-file-stub"
          />
          <span
            style={
              {
                "border": "0",
                "clip": "rect(0 0 0 0)",
                "height": "1px",
                "margin": "-1px",
                "overflow": "hidden",
                "padding": "0",
                "position": "absolute",
                "whiteSpace": "nowrap",
                "width": "1px",
              }
            }
          >
            More information
          </span>
        </button>
      </span>
    </div>
  </div>
  <div
    className="CopyNumberVariantPage__Wrapper-sc-1ase2rg-0 elCRDU"
  >
    <section
      className="CopyNumberVariantPage__ResponsiveSection-sc-1ase2rg-1 lfHhrQ"
    >
      <dl
        className="AttributeList-sc-1yjjegy-0 boDRvX"
        style={
          {
            "marginTop": "1.25em",
          }
        }
      >
        <div>
          <dt>
            Filter
          </dt>
          <dd>
            <span
              className="Badge__BadgeWrapper-sc-j4izdp-1 biZYYa"
            >
              Pass
            </span>
          </dd>
        </div>
        <div>
          <dt>
            Site Count
          </dt>
          <dd>
            123
          </dd>
        </div>
        <div>
          <dt>
            Site Number
          </dt>
          <dd>
            345
          </dd>
        </div>
        <div>
          <dt>
            Site Frequency
          </dt>
          <dd>
            0.3565
          </dd>
        </div>
        <div>
          <dt>
            Position
          </dt>
          <dd>
            <a
              className="Link-sc-14lgydv-0-Link jBvaYQ"
              href="/region/21-123-456"
              onClick={[Function]}
            >
              21
              :
              123
              -
              456
            </a>
          </dd>
        </div>
        <div>
          <dt>
            Size
          </dt>
          <dd>
            333 bp
          </dd>
        </div>
        <div>
          <dt>
            Class
          </dt>
          <dd>
            deletion
          </dd>
        </div>
      </dl>
    </section>
    <section
      className="CopyNumberVariantPage__ResponsiveSection-sc-1ase2rg-1 lfHhrQ"
    >
      <h2>
        External Resources
      </h2>
      <ul
        className="List-sc-4j87st-0 fQRmhM"
      >
        <li
          className="List__ListItem-sc-4j87st-1 gGPLUe"
        >
          <a
            className="Link-sc-14lgydv-0 Link__ExternalLink-sc-14lgydv-1 haNmsP"
            href="https://genome.ucsc.edu/cgi-bin/hgTracks?db=hg38&position=chr21%3A123-456"
            rel="noopener noreferrer"
            target="_blank"
          >
            UCSC
          </a>
        </li>
      </ul>
      <h2>
        Feedback
      </h2>
      <a
        className="Link-sc-14lgydv-0 Link__ExternalLink-sc-14lgydv-1 haNmsP"
        href="mailto:gnomad@broadinstitute.org"
        rel="noopener noreferrer"
        target="_blank"
      >
        Report an issue with this variant
      </a>
    </section>
  </div>
  <section>
    <h2>
      Genetic Ancestry Group Frequencies
    </h2>
    <table
      className="Table__BaseTable-sc-7fgtt2-0 CNVPopulationsTable__Table-sc-1dnyyn-0 iypcwR"
    >
      <thead>
        <tr>
          <th
            aria-sort="none"
            colSpan={2}
            scope="col"
          >
            <button
              onClick={[Function]}
              type="button"
            >
              Genetic ancestry group
            </button>
          </th>
          <th
            aria-sort="none"
            className="right-align"
            scope="col"
          >
            <button
              onClick={[Function]}
              type="button"
            >
              <span
                className="TooltipHint-sc-1wwbc8j-0 cXXQWi"
                onMouseEnter={[Function]}
                onMouseLeave={[Function]}
              >
                Site Count
              </span>
            </button>
          </th>
          <th
            aria-sort="none"
            className="right-align"
            scope="col"
          >
            <button
              onClick={[Function]}
              type="button"
            >
              <span
                className="TooltipHint-sc-1wwbc8j-0 cXXQWi"
                onMouseEnter={[Function]}
                onMouseLeave={[Function]}
              >
                Site Number
              </span>
            </button>
          </th>
          <th
            aria-sort="descending"
            className="right-align"
            scope="col"
          >
            <button
              onClick={[Function]}
              type="button"
            >
              <span
                className="TooltipHint-sc-1wwbc8j-0 cXXQWi"
                onMouseEnter={[Function]}
                onMouseLeave={[Function]}
              >
                Site Frequency
              </span>
            </button>
          </th>
        </tr>
      </thead>
      <tfoot>
        <tr
          className="border"
        >
          <th
            colSpan={2}
            scope="row"
          >
            Total
          </th>
          <td
            className="right-align"
          >
            123
          </td>
          <td
            className="right-align"
          >
            345
          </td>
          <td
            className="right-align"
          >
            0.3565
          </td>
        </tr>
      </tfoot>
    </table>
  </section>
  <div
    className="CopyNumberVariantPage__Wrapper-sc-1ase2rg-0 elCRDU"
  >
    <section
      className="CopyNumberVariantPage__ResponsiveSection-sc-1ase2rg-1 lfHhrQ"
    >
      <h2>
        Consequences
      </h2>
      <p>
        This variant has consequences in 
        0
         gene
        s
        .
      </p>
    </section>
  </div>
</div>
`;

exports[`CopyNumberVariantPage with dataset gnomad_cnv_r4 with variant of type DUP has no unexpected changes 1`] = `
<div
  className="Page-sc-w7h773-0 kodnkv"
>
  <div
    className="GnomadPageHeading__PageHeadingWrapper-sc-sgz3rr-0 gcOcdP"
  >
    <div
      className="GnomadPageHeading__PageHeadingInnerWrapper-sc-sgz3rr-1 dLZzAa"
    >
      <h1
        className="GnomadPageHeading__PageHeadingText-sc-sgz3rr-3 edXBqX"
      >
        Copy number variant: 
        21__DUP
      </h1>
    </div>
    <div
      className="GnomadPageHeading__PageControlsWrapper-sc-sgz3rr-4 liqJpb"
    >
      <span
        className="GnomadPageHeading__Label-sc-sgz3rr-5 dOhShJ"
      >
        Dataset
      </span>
      <ul
        className="DatasetSelector__NavigationMenuWrapper-sc-1p2fxbn-0 ebZvwy"
      >
        <li
          className="DatasetSelector__NavigationMenuItem-sc-1p2fxbn-2 ifirRc"
        >
          <a
            className="DatasetSelector__TopLevelNavigationLink-sc-1p2fxbn-1-Link eggTrD"
            data-item="current_sv_dataset"
            href="/?dataset=gnomad_cnv_r4"
            onBlur={[Function]}
            onClick={[Function]}
            onKeyDown={[Function]}
          >
            gnomAD CNVs v4.0
          </a>
        </li>
        <li
          className="DatasetSelector__NavigationMenuItem-sc-1p2fxbn-2 ifirRc"
        >
          <a
            aria-expanded={false}
            aria-haspopup="true"
            aria-label="More datasets"
            className="DatasetSelector__TopLevelNavigationLink-sc-1p2fxbn-1-Link eggTrD"
            data-item="other_structural_variant"
            href="/"
            onBlur={[Function]}
            onClick={[Function]}
            onKeyDown={[Function]}
            role="button"
          >
            <span
              className="DatasetSelector__MoreIcon-sc-1p2fxbn-3 hTYFQu"
            >
              <img
                alt=""
                aria-hidden="true"
                src="test-file-stub"
              />
            </span>
          </a>
          <ul
            className="DatasetSelector__SubNavigationMenu-sc-1p2fxbn-4 epYyTQ"
          >
            <li>
              <div
                className="DatasetSelector__GroupedNav-sc-1p2fxbn-7 cJJZPz"
              >
                GRCh38
              </div>
              <li>
                <a
                  className="DatasetSelector__SubNavigationLink-sc-1p2fxbn-5-Link gbnrKe"
                  data-item="gnomad_sv_r4"
                  href="/?dataset=gnomad_sv_r4"
                  onBlur={[Function]}
                  onClick={[Function]}
                  onKeyDown={[Function]}
                >
                  gnomAD SVs v4.0
                  <div
                    className="DatasetSelector__ItemDescription-sc-1p2fxbn-6 jjgqnf"
                  >
                    63,046 samples, genome
                  </div>
                </a>
              </li>
              <li>
                <a
                  className="DatasetSelector__SubNavigationLink-sc-1p2fxbn-5-Link gbnrKe"
                  data-item="gnomad_cnv_r4"
                  href="/?dataset=gnomad_cnv_r4"
                  onBlur={[Function]}
                  onClick={[Function]}
                  onKeyDown={[Function]}
                >
                  gnomAD CNVs v4.0
                  <div
                    className="DatasetSelector__ItemDescription-sc-1p2fxbn-6 jjgqnf"
                  >
                    464,297 samples, exome, rare (&lt;0.01)
                  </div>
                </a>
              </li>
            </li>
            <li>
              <div
                className="DatasetSelector__GroupedNav-sc-1p2fxbn-7 cJJZPz"
              >
                GRCh37
              </div>
              <li>
                <a
                  className="DatasetSelector__SubNavigationLink-sc-1p2fxbn-5-Link gbnrKe"
                  data-item="gnomad_sv_r2_1"
                  href="/?dataset=gnomad_sv_r2_1"
                  onBlur={[Function]}
                  onClick={[Function]}
                  onKeyDown={[Function]}
                >
                  gnomAD SVs v2.1
                  <div
                    className="DatasetSelector__ItemDescription-sc-1p2fxbn-6 jjgqnf"
                  >
                    10,847 samples, genome
                  </div>
                </a>
              </li>
              <li>
                <a
                  className="DatasetSelector__SubNavigationLink-sc-1p2fxbn-5-Link gbnrKe"
                  data-item="gnomad_sv_r2_1_non_neuro"
                  href="/?dataset=gnomad_sv_r2_1_non_neuro"
                  onBlur={[Function]}
                  onClick={[Function]}
                  onKeyDown={[Function]}
                >
                  gnomAD SVs v2.1 (non-neuro)
                  <div
                    className="DatasetSelector__ItemDescription-sc-1p2fxbn-6 jjgqnf"
                  >
                    8,342 samples, genome
                  </div>
                </a>
              </li>
              <li>
                <a
                  className="DatasetSelector__SubNavigationLink-sc-1p2fxbn-5-Link gbnrKe"
                  data-item="gnomad_sv_r2_1_controls"
                  href="/?dataset=gnomad_sv_r2_1_controls"
                  onBlur={[Function]}
                  onClick={[Function]}
                  onKeyDown={[Function]}
                >
                  gnomAD SVs v2.1 (controls)
                  <div
                    className="DatasetSelector__ItemDescription-sc-1p2fxbn-6 jjgqnf"
                  >
                    5,192 samples, genome
                  </div>
                </a>
              </li>
            </li>
          </ul>
        </li>
      </ul>
      <span>
        <button
          className="InfoButton__Button-sc-13t5e82-0 gxsoyZ"
          onClick={[Function]}
          type="button"
        >
          <img
            alt=""
            aria-hidden="true"
            src="test-file-stub"
          />
          <span
            style={
              {
                "border": "0",
                "clip": "rect(0 0 0 0)",
                "height": "1px",
                "margin": "-1px",
                "overflow": "hidden",
                "padding": "0",
                "position": "absolute",
                "whiteSpace": "nowrap",
                "width": "1px",
              }
            }
          >
            More information
          </span>
        </button>
      </span>
    </div>
  </div>
  <div
    className="CopyNumberVariantPage__Wrapper-sc-1ase2rg-0 elCRDU"
  >
    <section
      className="CopyNumberVariantPage__ResponsiveSection-sc-1ase2rg-1 lfHhrQ"
    >
      <dl
        className="AttributeList-sc-1yjjegy-0 boDRvX"
        style={
          {
            "marginTop": "1.25em",
          }
        }
      >
        <div>
          <dt>
            Filter
          </dt>
          <dd>
            <span
              className="Badge__BadgeWrapper-sc-j4izdp-1 biZYYa"
            >
              Pass
            </span>
          </dd>
        </div>
        <div>
          <dt>
            Site Count
          </dt>
          <dd>
            123
          </dd>
        </div>
        <div>
          <dt>
            Site Number
          </dt>
          <dd>
            345
          </dd>
        </div>
        <div>
          <dt>
            Site Frequency
          </dt>
          <dd>
            0.3565
          </dd>
        </div>
        <div>
          <dt>
            Position
          </dt>
          <dd>
            <a
              className="Link-sc-14lgydv-0-Link jBvaYQ"
              href="/region/21-123-456"
              onClick={[Function]}
            >
              21
              :
              123
              -
              456
            </a>
          </dd>
        </div>
        <div>
          <dt>
            Size
          </dt>
          <dd>
            333 bp
          </dd>
        </div>
        <div>
          <dt>
            Class
          </dt>
          <dd>
            duplication
          </dd>
        </div>
      </dl>
    </section>
    <section
      className="CopyNumberVariantPage__ResponsiveSection-sc-1ase2rg-1 lfHhrQ"
    >
      <h2>
        External Resources
      </h2>
      <ul
        className="List-sc-4j87st-0 fQRmhM"
      >
        <li
          className="List__ListItem-sc-4j87st-1 gGPLUe"
        >
          <a
            className="Link-sc-14lgydv-0 Link__ExternalLink-sc-14lgydv-1 haNmsP"
            href="https://genome.ucsc.edu/cgi-bin/hgTracks?db=hg38&position=chr21%3A123-456"
            rel="noopener noreferrer"
            target="_blank"
          >
            UCSC
          </a>
        </li>
      </ul>
      <h2>
        Feedback
      </h2>
      <a
        className="Link-sc-14lgydv-0 Link__ExternalLink-sc-14lgydv-1 haNmsP"
        href="mailto:gnomad@broadinstitute.org"
        rel="noopener noreferrer"
        target="_blank"
      >
        Report an issue with this variant
      </a>
    </section>
  </div>
  <section>
    <h2>
      Genetic Ancestry Group Frequencies
    </h2>
    <table
      className="Table__BaseTable-sc-7fgtt2-0 CNVPopulationsTable__Table-sc-1dnyyn-0 iypcwR"
    >
      <thead>
        <tr>
          <th
            aria-sort="none"
            colSpan={2}
            scope="col"
          >
            <button
              onClick={[Function]}
              type="button"
            >
              Genetic ancestry group
            </button>
          </th>
          <th
            aria-sort="none"
            className="right-align"
            scope="col"
          >
            <button
              onClick={[Function]}
              type="button"
            >
              <span
                className="TooltipHint-sc-1wwbc8j-0 cXXQWi"
                onMouseEnter={[Function]}
                onMouseLeave={[Function]}
              >
                Site Count
              </span>
            </button>
          </th>
          <th
            aria-sort="none"
            className="right-align"
            scope="col"
          >
            <button
              onClick={[Function]}
              type="button"
            >
              <span
                className="TooltipHint-sc-1wwbc8j-0 cXXQWi"
                onMouseEnter={[Function]}
                onMouseLeave={[Function]}
              >
                Site Number
              </span>
            </button>
          </th>
          <th
            aria-sort="descending"
            className="right-align"
            scope="col"
          >
            <button
              onClick={[Function]}
              type="button"
            >
              <span
                className="TooltipHint-sc-1wwbc8j-0 cXXQWi"
                onMouseEnter={[Function]}
                onMouseLeave={[Function]}
              >
                Site Frequency
              </span>
            </button>
          </th>
        </tr>
      </thead>
      <tfoot>
        <tr
          className="border"
        >
          <th
            colSpan={2}
            scope="row"
          >
            Total
          </th>
          <td
            className="right-align"
          >
            123
          </td>
          <td
            className="right-align"
          >
            345
          </td>
          <td
            className="right-align"
          >
            0.3565
          </td>
        </tr>
      </tfoot>
    </table>
  </section>
  <div
    className="CopyNumberVariantPage__Wrapper-sc-1ase2rg-0 elCRDU"
  >
    <section
      className="CopyNumberVariantPage__ResponsiveSection-sc-1ase2rg-1 lfHhrQ"
    >
      <h2>
        Consequences
      </h2>
      <p>
        This variant has consequences in 
        0
         gene
        s
        .
      </p>
    </section>
  </div>
</div>
`;
